### PR TITLE
[property-access] Improvement for Accessing Reference Chain

### PR DESCRIPTION
**Pull request by zerustech**
Improve performance for the following scenarios:
- Example 1:

``` php
$a = array(
    'a' => array(
        'b' => array(
            'c' => 'old-value'
        )
    )
);
$pa->setValue($a, '[a][b][c]', 'new-value');
// The PropertyAccessor will try to set values for
// $a['a']['b']['c'], $a['a']['b'] and $a['a'],
// but in fact it may terminate the loop
// right after the value of $a[a][b][c] is set,
// because $a, $[a], $[a][b] and $[a][b][c]
// are all passed as reference - the reference chain is not broken.
```
- Example 2

``` php
$b = array(
    'a' => array(
        'b' => array(
            'c' => 'old-value'
        )
    )
)

$a = new Foo($b);
// In this example, the reference chain of $b is broken,
// because it's passed to $a.value as value
// But its elements are all passed as reference,
// so after setting the value for $b[a][b][c], there is no need
// to set value for $b[a][b] and $b[a]

$pa->setValue($a, 'value[a][b][c]', 'new-value');
```

| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | n/a |

### CODE REVIEW

**Comment by hacfi** [comment]
:+1: Much cleaner solution

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 1]
```diff
@@ -82,14 +84,34 @@ public function setValue(&$objectOrArray, $propertyPath, $value)
...
 
-            if ($propertyPath->isIndex($i)) {
-                $this->writeIndex($objectOrArray, $property, $value);
-            } else {
-                $this->writeProperty($objectOrArray, $property, $value);
-            }
+            // You only need set value for current element if:
+            // 1. it's the parent of the last index element
+            // OR
+            // 2. it's child is not passed by reference
```
Comment (Triiistan) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: its

Comment (zerustech) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: @Triiistan Fixed

**Comment by stof** [comment]
@webmozart can you check this please ?

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 2]
```diff
@@ -82,14 +84,34 @@ public function setValue(&$objectOrArray, $propertyPath, $value)
...
+            // 2. its child is not passed by reference
+            //
+            // This may avoid uncessary value setting process for array elements.
+            // For example:
+            // '[a][b][c]' => 'old-value'
+            // If you want to change its value to 'new-value',
+            // you only need set value for '[a][b][c]' and it's safe to ignore '[a][b]' and '[a]'
+            //
+            if ($i === count($propertyValues) - 1 || !$propertyValues[$i + 1][self::IS_REF]) {
+
```
Comment (fabpot) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: empty newline should be removed

**Comment by fabpot** [comment]
:+1:

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php (hunk 2 above)
Comment (zerustech) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: @fabpot Fixed

**Comment by fabpot** [comment]
Thank you @zerustech.

**Comment by hacfi** [comment]
Awesome! Thanks @zerustech for the improvement!

**Comment by zerustech** [comment]
@hacfi @fabpot Thanks for your support too.

**Pull request merged**